src/main/java/scala/compat/java8/runtime/LambdaDeserializer.scala

```diff
@@ -0,0 +1,135 @@
+package scala.compat.java8.runtime
+
+import java.lang.invoke._
+import java.lang.ref.WeakReference
+
+/**
+ * This class is only intended to be called by synthetic `$deserializeLambda$` method that the Scala 2.12
+ * compiler will add to classes hosting lambdas.
+ *
+ * It is intended to be consumed directly.
+ */
+object LambdaDeserializer {
+  private final case class CacheKey(implClass: Class[_], implMethodName: String, implMethodSignature: String)
+  private val cache = new java.util.WeakHashMap[CacheKey, WeakReference[CallSite]]()
```

Comment:
I don't think it should be bounded. But your comment prompted me to reconsider this cache, and I've found a problem with it: while the weak references to the keys prevents the classic problem of strong references from the value back to the weak key undermining the attempted weakness of the cache, it also makes the key eligible for garbage collection sooner than we would like.

https://github.com/retronym/scala-java8-compat/compare/topic/lambda-deserialize...retronym:topic/lambda-deserialize-weak

So rather than `WeakHashMap[CacheKey, WeakReference[CallSite]]`, we need something like `WeakHashMap[CacheKey, WeakReference[Class] => CallSite]`.